plugins/filter_aws/aws.c

```diff
@@ -427,8 +429,135 @@ static int get_ec2_tag_values(struct flb_filter_aws *ctx)
     return 0;
 }
 
+static int tag_is_present_in_list(struct flb_filter_aws *ctx, flb_sds_t tag,
+        flb_sds_t *tags, int tags_n)
+{
+    int i;
+    for (i = 0; i < tags_n; i++) {
+        if (strcmp(tag, tags[i]) == 0) {
+            return FLB_TRUE;
+        }
+    }
+    return FLB_FALSE;
+}
+
+static int tags_split(char *tags, flb_sds_t **tags_list, int *tags_list_n) {
+    flb_sds_t token;
+    int i;
+    int n;
+    n = 1;
+    for (i = 0; i < strlen(tags); i++) {
+        if (tags[i] == ',') {
+            n++;
+        }
+    }
+
+    *tags_list = calloc(sizeof(flb_sds_t), n);
+    if (*tags_list == NULL) {
+        return -2;
+    }
+
+    token = strtok(tags, ",");
+    i = 0;
+    while (token != NULL) {
+        (*tags_list)[i] = token;
+        i++;
+        token = strtok(NULL, ",");
+    }
+
+    *tags_list_n = n;
+
+    return 0;
+}
+
+static int get_ec2_tag_enabled(struct flb_filter_aws *ctx)
+{
+    const char *tags_include;
+    const char *tags_exclude;
+    char *tags_copy;
+    flb_sds_t *tags;
+    int tags_n;
+    int i;
+    int tag_present;
+    int result;
+
+    /* if there are no tags, there is no need to evaluate which tag is enabled */
+    if (ctx->tags_count == 0) {
+        return 0;
+    }
+
+
+    /* allocate memory for 'tag_is_enabled' for all tags */
+    ctx->tag_is_enabled = flb_calloc(ctx->tags_count, sizeof(int));
+    if (!ctx->tag_is_enabled) {
+        flb_plg_error(ctx->ins, "Failed to allocate memory for tag_is_enabled");
+        return -1;
+    }
+
+    /* if tags_include and tags_exclude are not defined, set all tags as enabled */
+    for (i = 0; i < ctx->tags_count; i++) {
+        ctx->tag_is_enabled[i] = FLB_TRUE;
+    }
+
+    /* apply tags_included configuration */
+    tags_include = flb_filter_get_property("tags_include", ctx->ins);
+    if (tags_include) {
+        /* copy const string in order to use strtok which modifes the string */
+        tags_copy = flb_strdup(tags_include);
+        if (!tags_copy) {
+            return -1;
+        }
+        result = tags_split(tags_copy, &tags, &tags_n);
+        if (result < 0) {
+            free(tags_copy);
+            return result;
+        }
+        for (i = 0; i < ctx->tags_count; i++) {
+            tag_present = tag_is_present_in_list(ctx, ctx->tag_keys[i], tags, tags_n);
+            /* tag is enabled if present in included list */
+            ctx->tag_is_enabled[i] = tag_present;
+        }
+        free(tags_copy);
+        free(tags);
+    }
+
+    /* apply tags_excluded configuration, only if tags_included is not defined */
+    tags_exclude = flb_filter_get_property("tags_exclude", ctx->ins);
+    if (tags_include && tags_exclude) {
+        flb_plg_warn(ctx->ins, "specyfing tags_include and tags_exclude at the same time"
```

Comment:
Normally in FLB we always fail the plugin registration if the config is invalid. I think we should do that here too, unless you have an argument against it.

Reply:
I agree. If that's the expected behaviour of the FLB plugins, then we should do the same here. I will try to change this within 2 days.

Reply:
thanks

Reply:
I am not sure if this is explained well enough. If I wouldn't know how it works underneath, I might have not been able to fix it.

What do you think about changing the log level of 'specyfing tags_include and tags_exclude...' with an error level?

Reply:
Yea that definitely needs to be an error

Reply:
https://github.com/fluent/fluent-bit/pull/6866/commits/ba85ff53e7c1971d97cc8438fbd5e9b69cd9c406

```
mwarzynski@57af70bc2a04:/fluent-bit-dev/build$ ./bin/fluent-bit -c ./fluent-bit.conf
Fluent Bit v2.1.0
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/03/19 12:13:31] [ info] [fluent bit] version=2.1.0, commit=8d937cfb6a, pid=17811
[2023/03/19 12:13:31] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/03/19 12:13:31] [ info] [cmetrics] version=0.5.8
[2023/03/19 12:13:31] [ info] [ctraces ] version=0.3.0
[2023/03/19 12:13:31] [ info] [input:stdin:stdin.0] initializing
[2023/03/19 12:13:31] [ info] [input:stdin:stdin.0] storage_strategy='memory' (memory only)
[2023/03/19 12:13:31] [error] [filter:aws:aws.0] configuration is invalid, both tags_include and tags_exclude are specified at the same time
[2023/03/19 12:13:31] [error] [filter:aws:aws.0] Failed to get instance EC2 Tags
[2023/03/19 12:13:31] [error] Failed initialize filter aws.0
[2023/03/19 12:13:31] [error] [engine] filter initialization failed
[2023/03/19 12:13:31] [error] [lib] backend failed
```